Doc/regions.md

```diff
@@ -0,0 +1,517 @@
+Its reference count can be modified and it can be collected if it is no longer referenced.
+This reference count may be concurrently modified.
+
+To allow globals to be used in a concurrent setting, we add an explicit operation to make the globals immutable.
+```
+make_globals_immutable()
+```
+This must have been run to start the concurrency runtime and enable region creation.
+This cannot be immediately done as we need some time to allow for the set-up of the types and any monkey-patching.
+As types are globals, you would not be able to have any fields on Types (for example default values) if they are immutable from the start.
+
+We do not expect the `make_globals_immutable()` operation to be exposed to the programmer directly, but to be called as needed by the runtime.
+Exposing it in the prototype is useful for testing and debugging.
+
+## Project plan
+
+We need to break the development in to several phases and their completion requirements.
+
+### Phase 1: Deeply immutable objects
+
+*  A flag to represent if an object is deeply immutable.
+*  If the flag is set, then the object cannot be mutated.  (This needs to be enforced in the runtime.)
+*  Any object reachable from a deeply immutable object is also deeply immutable.
+*  A `freeze` operation that makes an object and its reachable graph of objects deeply immutable. 
+*  Reference counts on immutable objects should be atomic instructions.
+
+### Phase 2: `make_globals_immutable`
+
+A function that makes all global objects deeply immutable.
+
+### Phase 3: Regions
+
+* Adding a field to the object header to represent the region.
+* A new object type to represent regions.
+* An operation to promote an object into a region.
+* This requires tracking the number of references from local objects to a region (as described above) called the "local reference count", LRC.
+
+### Phase 4: Write barrier
+
+A write-barrier that ensures the region's local reference count is maintained.
+
+The following cases (that may overlap) need to be handled:
+
+* Changing a reference from a "local object" from into a "region", then we need to remove a reference from the LRC.
+* Changing a reference from a "local object" to point into a "region", then we need to add a reference to the LRC.
+* Creating a reference from an object in a "region" to a "local object", then we need to promote the local object to also be in that region.
+  This may fail if the local object can reach a different region. 
+
+### Phase 5: Send/consume
+
+Add an operation that represents sending a region to another thread.
+This is a placeholder operation that will be turned into proper concurrency support later in Phase 7.
+
+If the region still has a non-zero LRC, then we need to 
+
+* scan the current frame to find all references to the region, and remove them.
+* If there are still references, then scan all objects (from the cycle detector DLL) to find a reference, and report that as an error.
+
+### Phase 6: Memory management
```

Comment:
Does it make sense to use the SCC stuff for immutables here as well? If we have logic to detect that an RC manipulation should be atomic because an object is immutable, this seems like an easy addition.